editor/scene_tree_dock.cpp

```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+
+			for (Node *F : owned) {
+				F->set_owner(p_owner);
+			}
+		}
+	}
+}
+
+void SceneTreeDock::_reparent_nodes_to_paths_with_transform_and_name(Node *p_root, Array p_nodes, Array p_paths, Array p_transforms, Array p_names, Node *p_owner) {
+	ERR_FAIL_COND(p_nodes.size() != p_paths.size());
+	ERR_FAIL_COND(p_nodes.size() != p_transforms.size());
+	ERR_FAIL_COND(p_nodes.size() != p_names.size());
+
+	for (int i = 0; i < p_nodes.size(); i++) {
+		Node *node = Object::cast_to<Node>(p_nodes[i]);
+		ERR_FAIL_COND(node == nullptr);
+		NodePath np = p_paths[i];
+		Node *parent_node = p_root->get_node_or_null(np);
+		ERR_FAIL_COND(!parent_node);
+
+		List<Node *> owned;
+		node->get_owned_by(p_owner, &owned, true);
+		node->reparent(parent_node);
+		node->set_name(p_names[i]);
+
+		for (Node *F : owned) {
+			F->set_owner(p_owner);
+		}
+	}
+}
+
 void SceneTreeDock::_toggle_editable_children(Node *p_node) {
 	if (p_node) {
+		EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+
+		undo_redo->create_action(TTR("Toggle Editable Children"));
+
 		bool editable = !EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(p_node);
-		EditorNode::get_singleton()->get_edited_scene()->set_editable_instance(p_node, editable);
+
+		undo_redo->add_undo_method(EditorNode::get_singleton()->get_edited_scene(), "set_editable_instance", p_node, !editable);
+		undo_redo->add_do_method(EditorNode::get_singleton()->get_edited_scene(), "set_editable_instance", p_node, editable);
+
 		if (editable) {
-			p_node->set_scene_instance_load_placeholder(false);
+			bool original_scene_instance_load_placeholder = p_node->get_scene_instance_load_placeholder();
+
+			undo_redo->add_undo_method(p_node, "set_scene_instance_load_placeholder", original_scene_instance_load_placeholder);
+			undo_redo->add_do_method(p_node, "set_scene_instance_load_placeholder", false);
+		} else {
+			Node *edited_scene = EditorNode::get_singleton()->get_edited_scene();
+
+			List<Node *> owned;
+			p_node->get_owned_by(edited_scene, &owned, true);
+
+			// Get the original paths, transforms, and names for undo
+			Array owned_nodes_array;
+			Array paths_array;
+			Array transform_array;
+			Array name_array;
+
+			for (Node *owned_node : owned) {
+				owned_nodes_array.push_back(owned_node);
+				paths_array.push_back(p_node->get_path_to(owned_node->get_parent()));
+				name_array.push_back(owned_node->get_name());
+				Node3D *node_3d = Object::cast_to<Node3D>(owned_node);
+				if (node_3d) {
+					transform_array.push_back(node_3d->get_transform());
+				} else {
+					Node2D *node_2d = Object::cast_to<Node2D>(owned_node);
+					if (node_2d) {
+						transform_array.push_back(node_2d->get_transform());
+					} else {
+						transform_array.push_back(Variant());
+					}
+				}
+			}
+
+			undo_redo->add_undo_method(this, "_reparent_nodes_to_paths_with_transform_and_name", p_node, owned_nodes_array, paths_array, transform_array, name_array, edited_scene);
```

Comment:
I think this method and the one below could be merged. You don't need transform and name arrays, as noted in comments above, and paths can be empty when reparenting to root. Though maybe I'm missing something.

Reply:
While it seems overkill, I can explain why we have the names and transforms array. For explanation of names, see the explanation to the above response about undoing naming conflicts. As for the transforms, while its uncommon given a revert chain should put things back to its relative position before the reparent, storing the original transforms will restore it in situations where the local position might change. In example I was testing, I had a scene with an animated skeleton and bone attachments. If I instance this scene in another scene, make it editable, then attach a mesh to that bone attachment, then remove the editable status, the mesh will become detached. This is fine, but if the skeleton is animating, we don't know how to restore it local position by just using reparenting, hence the use of the transform array.

Admittedly, yeah, it this wasn't doing anything because the code to restore the local transform got lost during the migration to the reparent function, but I've restored it now.